processing/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -1188,11 +1236,60 @@ boolean isCancelled()
 
     RuntimeException getRuntimeException()
     {
-      Throwable ex = throwable.get();
-      if (ex instanceof RuntimeException) {
-        return (RuntimeException) ex;
+      return wrapRuntimeException(throwable.get());
+    }
+
+    private static RuntimeException wrapRuntimeException(Throwable t)
+    {
+      if (t instanceof RuntimeException) {
+        return (RuntimeException) t;
       }
-      return new RE(ex);
+      return new RuntimeException(t);
+    }
+  }
+
+  /**
+   * {@link com.google.common.util.concurrent.ListenableFuture} that allows {@link ParallelMergeCombiningSequence} to be
+   * registered with {@link org.apache.druid.query.QueryWatcher#registerQueryFuture} to participate in query
+   * cancellation or anything else that has a need to watch the activity on the merge pool. Wraps a
+   * {@link CancellationGizmo} to allow for external threads to signal cancellation of parallel processing on the pool
+   * by triggering {@link CancellationGizmo#cancel(Throwable)} whenever {@link #cancel(boolean)} is called.
+   *
+   * This is not used internally by workers on the pool in favor of using the much simpler {@link CancellationGizmo}
+   * directly instead.
+   */
+  public static class CancellationFuture extends AbstractFuture<Boolean>
+  {
+    private final CancellationGizmo cancellationGizmo;
+
+    public CancellationFuture(CancellationGizmo cancellationGizmo)
+    {
+      this.cancellationGizmo = cancellationGizmo;
+    }
+
+    public CancellationGizmo getCancellationGizmo()
+    {
+      return cancellationGizmo;
+    }
+
+    @Override
+    public boolean set(Boolean value)
+    {
+      return super.set(value);
+    }
+
+    @Override
+    public boolean setException(Throwable throwable)
+    {
+      cancellationGizmo.cancel(throwable);
+      return super.setException(throwable);
+    }
+
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning)
+    {
+      cancellationGizmo.cancel(new RuntimeException("Sequence cancelled"));
```

Comment:
canceled (spelling)

We aren't super consistent about this, but I am advocating for being more consistent using this style: https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/cancel-canceled-canceling-cancellation. That came after doing a bunch of research when developing the cancellation features for the multi-stage-query engine. (I don't have the link to the discussion handy, though.)

Reply:
cool